### PR TITLE
KAFKA-8131: Moved --version argument implementation from kafka-run-class.sh to CommandLineUtils to include documentation in command line help

### PR DESCRIPTION
**Pull request by soenkeliebau**
After some further investigation, I've taken the liberty of refactoring the implementation of the --version option and moved it into the default command options. This has the benefit of automatically including it in the usage output of the command line tools without having to actually touch any of them.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by soenkeliebau** [comment]
I've not yet included documentation and tests for this refactored code, as I'd like to discuss whether this makes sense first.

**Comment by soenkeliebau** [comment]
@hachikuji - you were one of the reviewers on the original addition of the parameter, could I perhaps trouble you to give this a quick once over?

**Comment by soenkeliebau** [comment]
@guozhangwang or @cmccabe maybe one of the two of you can have a quick look? ou I saw that both of you reviewed the addition of CommandDefaultOptions which I change in this...

**Review by guozhangwang — APPROVED** [review]
@soenkeliebau Thanks for the PR. Your approach looks clean and sound to me.

cc anyone of @ijuma @cmccabe @omkreddy @mumrah @gwenshap @rajinisivaram for another pair of eyes.

**Review by hachikuji — COMMENTED** [review]
@soenkeliebau Apologies for the delay. It looks good to me, but let me check again in the morning when I'm a bit more sober. 😉

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/utils/CommandLineUtils.scala]: nit: return is implicit in scala

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: Hmm.. I'm not sure this is called by all the commands that go through `kafka-run-class.sh`. For example, `kafka.Kafka` does not seem to hit this path.

**Review by hachikuji — COMMENTED** [review]
Thanks for the PR. Left a few comments.

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala
Comment (soenkeliebau) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala]: Just can't get used to that :)
I'll fix it.

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala
Comment (soenkeliebau) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: I've been through the scripts in bin/ and looked at how many would be affected by this.  There are three classes that use CommandLineUtils but don't hit this method:  ReplicaVerificationTool, Kafka, StreamsResetter. 
Coincidentally for StreamsResetter this means that the "--help" option, while listed in the output still results in an error.

The issue with these classes is, I think, that they internally do not use an object that extends CommandDefaultOptions - hence ignore those options completely.

The rest of the tools either call this method or handle argument parsing in a totally different way.

Not sure what the correct way forward is, to be honest, command line parsing seems a little all over the place in general, so the only place where this is truly global is indeed in kafka-run-class.sh, but that doesn't feel right either, as it is a completely undocumented on the command line that way.

Comment (hachikuji) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: If there only 3 exceptions, do you think it would be reasonable to implement `--version` separately for each of them? Perhaps in a follow-up we can consider how to refactor so that they all use `CommandDefaultOptions`. What do you think?

Comment (soenkeliebau) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: Hi @hachikuji 
I am happy to implement those three separately for now.

In the interest of completion, handling those three would leave the following classes that do not support the version option (or help for that matter unless they implemented that separately):

- ConnectDistributed - Doesn't use CommandLine
- ConnectStandalone - Doesn't use CommandLine
- ProducerPerformance - Uses argparse4j instead
- VerifiableConsumer - Uses argparse4j instead
- VerifiableProducer - Uses argparse4j instead
- QuorumPeerMain - Zookeeper class
- ZooKeeperMain - Zookeeper class

I'm not sure why a small subset of these uses argparse4j, I briefly looked around but couldn't find a larger scheme to adopt this over the CommandLineUtils, do you know anything more about that?

Should we extend these classes as well to honor the version flag? Should be simple-ish, but still requires adding an option to all parsers, unless we create the equivalent of the CommandDefaultOptions for argparse4j classes. But I'd say this is better suited for the follow-up PR you mentioned to standardize command line parsing.

So, my proposal woud be:
add very simple implementations to ConnectDistributed and ConnectStandalone as well, leave the rest be for now and look at the larger parsing picture in a follow up PR.

Fair?

Comment (hachikuji) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: Yeah, I'm not sure why they use argparse4j. Maybe just to make our lives harder. Your suggestion sounds fine to me. Would you mind filing JIRAs for the remaining work?

Comment (soenkeliebau) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: Happy to!
I'll kick off a thread on the mailing list first though, to ask around what people think we should do, personally, I'd like us to use just one command line parser, don't really care which one, but two is one too many for my liking.
But maybe there were reasons for the split.

Comment (hachikuji) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: I suspect there are no good reasons, but starting a discussion is a good idea. There is an ancient KIP which was attempting to consolidate these tools: https://cwiki.apache.org/confluence/display/KAFKA/KIP-14+-+Tools+Standardization. If you have any interest, it would really be helpful for someone to pick that up again. I think one of the reasons for the naming inconsistencies has been the fact that we are using different libraries.

Comment (soenkeliebau) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:59]: Having looked some more at the two Connect classes, there is a small catch: everything we'd need to add the --version parameter to those classes is in the core Kafka project which Connect currently does not declare as a dependency. It seems a bit excessive to add that just for this purpose. 
So I'll reword my suggestion to ignore these two for now and come up with an overall action plan first, before making any larger changes.

File: core/src/main/scala/kafka/Kafka.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/Kafka.scala:47]: Should we do the same thing for `--help`?

File: core/src/main/scala/kafka/Kafka.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/Kafka.scala]: nit: if we don't need the variable, maybe we can just remove it?

For some reason, this wasn't working for me. I looked in the `kafka-server-start.sh` script and found the following:

```
if [ $# -lt 1 ];
then
        echo "USAGE: $0 [-daemon] server.properties [--override property=value]*"
        exit 1
fi
```

Because of this check, we cannot get the proper help message if we don't specify any arguments. It seems we can just get rid of it.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: Probably not a big deal, but we could be explicit and say "Print version information and exit."

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala]: Don't forget this 🙂

**Review by hachikuji — COMMENTED** [review]
Just a few additional comments

File: core/src/main/scala/kafka/Kafka.scala
Comment (soenkeliebau) [reply core/src/main/scala/kafka/Kafka.scala]: Hi Jason, I noticed that as well and opened [KAFKA-8247](https://issues.apache.org/jira/browse/KAFKA-8247) for it, as this seemed to me to broaden the scope of this PR too much beyond the related ticket.
I don't have any real preference though, happy to address it here as well.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala
Comment (soenkeliebau) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: Sure, I'll change it.

File: core/src/main/scala/kafka/Kafka.scala
Comment (soenkeliebau) [reply core/src/main/scala/kafka/Kafka.scala:47]: I've added an _or_ed option along the lines of what other classes do.
Should I extract that change into a different PR, it is not strictly speaking "in scope" for this one..

File: core/src/main/scala/kafka/Kafka.scala
Comment (soenkeliebau) [reply core/src/main/scala/kafka/Kafka.scala]: If we remove the variable, the --version flag won't show up in the _usage_ output, once we fix the duplicate (triplicate actually) parameter check. So it is not strictly speaking unused, just unused by our code, the argument parsing library looks at it.
If we decide to keep it we should also add the help parameter though. I've not changed anything yet, will either remove this or add help after we've discussed this. 
I'd prefer adding help, but that is not a strong preference.

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala
Comment (soenkeliebau) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala]: I indeed forgot :)
Done now.
I've also opened #6629 to remove the existing return in the code and not "pollute" this PR with that.

File: core/src/main/scala/kafka/Kafka.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/Kafka.scala]: Thanks, I hadn't noticed KAFKA-8247, but I think it makes sense to address this there.

**Review by hachikuji — APPROVED** [review]
Thanks for the patch. LGTM. Before I merge, can you point me to the follow-up JIRAs? I'm (now) aware of https://issues.apache.org/jira/browse/KAFKA-8247. I think we also need to add --version to a few of the remaining tools like Connect?

**Comment by soenkeliebau** [comment]
Thanks for reviewing @hachikuji !
I've created KAFKA-8292 for the remaining five scripts and stared a discussion thread on the mailing list about reworking the entire parsing thing a little.
Is that sufficient?

**Comment by soenkeliebau** [comment]
ping @hachikuji

**Comment by hachikuji** [comment]
@soenkeliebau Thanks for the reminder. This one got lost in the noise. Let me run tests again to be safe and then I will merge.

**Comment by hachikuji** [comment]
retest this please

**Pull request merged**
